### PR TITLE
[NFC] [DSE] Refactor DSE

### PR DESCRIPTION
**Pull request by haopliu**
Refactor DSE with MemoryDefWrapper and MemoryLocationWrapper.

Normally, one MemoryDef accesses one MemoryLocation. With "initializes" attribute, one MemoryDef (like call instruction) could initialize multiple MemoryLocations.

Refactor DSE as a preparation to apply "initializes" attribute in DSE in a follow-up PR (https://github.com/llvm/llvm-project/commit/58dd8a440343055b1a4929d72317218e912c16fd).

### CODE REVIEW

**Comment by tschuett** [comment]
Can promote somebody to code owner to make the assignment automatic?

**Comment by aeubanks** [comment]
with just this patch, it's still one MemoryLocation per MemoryDef. can we keep that assumption for now in this PR? and then make the change in the initializes DSE patch

**Comment by aeubanks** [comment]
also, the dependency cycle between Memory*Wrapper and DSEState is awkward, is it possible to keep the `eliminateDeadDefs()` implementations as part of DSEState?

**Comment by haopliu** [comment]
> with just this patch, it's still one MemoryLocation per MemoryDef. can we keep that assumption for now in this PR? and then make the change in the initializes DSE patch

It's correct! With this PR, we still keep one MemoryLocation per MemoryDef.
See `assert(DefinedLocations.size() == 1 && "Expected a single defined location");`
This PR only refactors and aims to be equivalent to the existing implementation.

> also, the dependency cycle between Memory*Wrapper and DSEState is awkward, is it possible to keep the `eliminateDeadDefs()` implementations as part of DSEState?

We can do that to avoid the dependency cycle. DSEState is a big struct (1300 lines of code). Is it ok to continue adding code in this struct?

**Comment by aeubanks** [comment]
> > with just this patch, it's still one MemoryLocation per MemoryDef. can we keep that assumption for now in this PR? and then make the change in the initializes DSE patch
> 
> It's correct! With this PR, we still keep one MemoryLocation per MemoryDef. See `assert(DefinedLocations.size() == 1 && "Expected a single defined location");` This PR only refactors and aims to be equivalent to the existing implementation.

I mean in this PR `DefinedLocation` shouldn't be a `SmallVector`, just a `MemoryLocationWrapper`. The PR that introduces multiple MemoryLocations per MemoryDef should make that change instead.
> 
> > also, the dependency cycle between Memory*Wrapper and DSEState is awkward, is it possible to keep the `eliminateDeadDefs()` implementations as part of DSEState?
> 
> We can do that to avoid the dependency cycle. DSEState is a big struct (1300 lines of code). Is it ok to continue adding code in this struct?

I thought the code was originally in DSEState?

**Comment by haopliu** [comment]
> I mean in this PR `DefinedLocation` shouldn't be a `SmallVector`, just a `MemoryLocationWrapper`. The PR that introduces multiple MemoryLocations per MemoryDef should make that change instead.

Changed the `DefinedLocations (SmallVector)` field to `DefinedLocation (optional)`.

> I thought the code was originally in DSEState?

The major code was originally in `static bool eliminateDeadStores(...)`, this loop:
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp#L2145

**Comment by haopliu** [comment]
> Moving a function from a static function to a DSEState method seems fine, I don't see anything wrong with that

Okay, moved MemoryDefWrapper and MemoryLocationWrapper to DSEState!

**Comment by aeubanks** [comment]
sorry, I meant that we should keep Memory*Wrapper as simple structs that just hold some data without any complex methods like `eliminateDeadDefs`, and then DSEState contains all the real logic, using Memory*Wrapper wherever necessary. not to move Memory*Wrapper under DSEState. Memory*Wrapper should probably be outside DSEState like before

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (aeubanks) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2192]: nit: this is returning a bool instead of a change state

**Review by aeubanks — COMMENTED** [review]
yes this is what I was looking for, thanks

I'll review this more in depth soon

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (aeubanks) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2189]: a lot of the enum values of `ChangeStateEnum` aren't actually checked. following the original code better, can we return a `std::pair<bool, bool>` of `[Changed, Shortened]`?

**Comment by aeubanks** [comment]
I'd add `[NFC]` to the commit title to emphasize this is pure refactoring

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Make sense. Done.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (aeubanks) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2181]: IIUC, `Shortened` really means "KillingI was deleted" (please correct me if I'm wrong). So maybe change the variable name to `DeletedKillingLoc` and the comment to

```
Try to eliminate dead defs killed by `KillingLocWrapper`. Return whether any changes were made, and whether `KillingLocWrapper` was deleted.
```

**Review by aeubanks — COMMENTED** [review]
aside from one comment, lgtm. do you have a followup DSE commit you could reference in this commit's description?

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (jvoung) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:838]: In a later stage this becomes a SmallVector of MemoryLocationWrapper. Is that right?

Do the MemoryLocationWrapper all have the same MemoryDef and DefInst?
I wonder if that's actually worth bundling in MemoryLocationWrapper (and thus repeating), or if the SmallVector should just be the MemoryLocation and UnderlyingObject (the non-repeating parts)?

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (jvoung) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2181]: nit:  "eliminate dead defs killed by `KillingLocWrapper`" I'm not sure killed by KillingLocWrapper is as clear. Which part of KillingLocWrapper are the defs killed by? It's not the MemoryLocation parts right?

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (jvoung) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2315]: Maybe naive question, but I'm having a bit of trouble determining where `MemDef` and `DefInst` are coming from in this scope. Are they fields of DSEState? How do they relate to the  KillingDefWrapper that is being processed?

**Review by jvoung — COMMENTED** [review]
Sorry for the delay in reviewing! A few comments

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Done!

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Inlined this function.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Ah, nice catch!

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Renamed "KillingLoc" parameter to "KillingLocWrapper", and updated the comments in `getDomMemoryDef` and two `eliminateDeadDefs` functions:
"KillingLoc" --> "KillingLocWrapper.MemLoc"
"KillingDef" --> "KillingLocWrapper.MemDef"

Let me know if we prefer to make a local var `KillingDef = KillingLocWrapper.MemDef`.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Done.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Done.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:836]: Yes, this will become a SmallVector of `MemoryLocationWrapper` and all have the same MemoryDef and DefInst.

Initially, I bundled them in `MemoryDefWrapper` but I realized that most accesses are through `MemoryLocationWrapper` in `getDomMemoryDef`. It seems that we should either bundle in `MemoryLocationWrapper` or keep a `MemoryDefWrapper` reference in each `MemoryLocationWrapper`. What do you think?

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Updated to "Try to eliminate dead defs that access `KillingLocWrapper.MemLoc` and are killed by `KillingLocWrapper.MemDef`".

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (jvoung) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:836]: Hmm my preference is not to keep a reference to `MemoryDefWrapper` in each `MemoryLocationWrapper` and not make it circular =)

Sorry, maybe I should have discussed this more before asking about the comment updates, but I didn't notice this until later (already added note about comments).

What do you think of splitting out the MemDef as a separate parameter to `getDomMemoryDef` as it was before (KillingDef)? And then DefInst can be derived in a local scope as before (KillingI). Does DefInst need to be extracted early?

Can you clarify the benefits of bundling them in a single parameter?
* Initially you had eliminateDeadDefs as a method of the wrapper, but that's now moved back into DSEState.
* I can see that there's a lot of parameters for getDomMemoryDef and the relationships are loose and only tied together by the same name prefix "Killing"
* similar for the "DeadLoc", "DeadDefAccess",  ...

But then
* there is the repetition
* the "Wrapper" part of `MemoryLocationWrapper` name doesn't convey much information to the reader (related to the "killed by" comments where it was unclear there's a MemDef in the MemoryLocationWrapper class, and now say "killed by `KillingLocWrapper.MemDef"). Similar for the MemoryDefWrapper but the other way -- it may be unclear there are MemoryLocations in the class. If both Wrappers have access to the Def and some Locations it feels harder to tell the difference from the name.

A middle ground could be a single class "MemDefAndLocations" for bundling purposes (Killing vs Dead, etc.), but then allow splitting off MemDef and MemoryLocation when the execution starts focusing on one of the MemoryLocations related to the MemDef?

Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:836]: > I can see that there's a lot of parameters for getDomMemoryDef and the relationships are loose and only tied together by the same name prefix "Killing"

This is the main reason. I tried to make `getDomMemoryDef` with a bit clearer interface: given a (killing) MemoryLocWrapper, return a MemoryDef that dominates the MemoryLoc or underlying object. This process needs DefInst but mostly for additional checks, I think, so I choose `MemoryLocWrapper` as a parameter and try to bundle all killing-related stuff in this parameter.

Thanks for raising this concern! I agree this bundling is too surface. May be an ideal way is to refactor/divide the `getDomMemoryDef` logic then bundle the members and the corresponding logic together.

How about we step back and keep `getDomMemoryDef` as it was?

Comment (jvoung) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:836]: > How about we step back and keep getDomMemoryDef as it was?

Sounds good, thanks!

Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:836]: Done. Please take another look. Thanks!

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:812]:
```suggestion
struct MemoryLocationWrapper {
```

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:829]:
```suggestion
struct MemoryDefWrapper {
```

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2179]: Can KillingLocWrapper be modified? If so, document, if not make const?

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2183]: Can KillingLocWrapper be modified? If so, document, if not make const?

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Done, thanks!

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Done!

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Marked with const!

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp
Comment (haopliu) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp]: Marked with const!

**Review by jvoung — COMMENTED** [review]
small nit but otherwise lgtm too

**Comment by haopliu** [comment]
Thank you all! Here is the follow-up DSE commit for reference (https://github.com/llvm/llvm-project/commit/58dd8a440343055b1a4929d72317218e912c16fd). 
Will merge this refactoring PR this week if no other comments.

**Review by aeubanks — APPROVED** [review]
lgtm

**Pull request merged**
